eth/abc.py

```diff
@@ -2269,6 +2270,59 @@ def get_transaction_context(cls,
         ...
 
 
+class ConsensusContextAPI(ABC):
+    """
+    A class representing a data context for the :class:`~eth.abc.ConsensusAPI` that remains static
+    over the life time of a chain instance.
+    """
+
+    @abstractmethod
+    def __init__(self, db: AtomicDatabaseAPI) -> None:
+        """
+        Initialize the context with a database.
```

Comment:
Couple random thoughts, as I won't be able to pick up the review again until later:

I was assuming we would formalize the idea that this context can't write to the database (Maybe this should be a read-only version that's supplied?) Or if we don't then maybe it fits as a kind of parallel to `ChainDB`, like `ConsensusDB`.

"remains static" in the docs above didn't immediately mean to me what I think we want to say. Something like: this instance stays in memory across VM runs.

Reply:
>I was assuming we would formalize the idea that this context can't write to the database 

But, we need to write to the database. E.g. Clique persists snapshots to the db.

https://github.com/ethereum/py-evm/blob/7b94eb17d43d08ac7732a0d6f935190423b5e33e/eth/consensus/clique/snapshot_manager.py#L248

>Or if we don't then maybe it fits as a kind of parallel to `ChainDB`, like `ConsensusDB`

I'm not sure how that would work out. In the `ChainDB` case, we know the specific types that make a chain and that can be stored and retrieved. But for consensus, it seems pretty much up to the specific consensus algorithm which kind of things need persistence. Clique uses snapshots but it's hard to imagine all the different things other consensus schemes would require.

>"remains static" in the docs above didn't immediately mean to me what I think we want to say. Something like: this instance stays in memory across VM runs.

:+1: I'll update that.